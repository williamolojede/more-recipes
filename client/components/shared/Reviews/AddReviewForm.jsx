import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { postReview } from '../../../actions/fetchRecipe';

import UserImg from '../UserImg.jsx';
import Preloader from '../Preloader';

/**
 * @param {*} props
 * @return {JSX} - Add review form
 */
class AddReviewForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      sendingreview: false
    };
  }
  addReview = () => {
    const content = this.reviewInput.value.trim();
    if (content) {
      this.setState({
        sendingreview: true,
      });
      this.props.dispatch(postReview(content, this.props.id))
        .then(() => {
          this.setState({
            sendingreview: false,
          });
        });
      this.reviewInput.value = '';
    }
  }
  render() {
    const { user } = this.props;

    return (
      <li className="review__card review__card-edit card" id="write-review">
        {this.state.sendingreview && <Preloader />}
        {!this.state.sendingreview &&
          <div className="card-image">
            <UserImg user={user} type="inReview" />
            <h5 className="card-title">{user.fullname}</h5>
          </div>
        }
        {!this.state.sendingreview &&
          <div className="card-content">
            <textarea placeholder="your review..." ref={node => this.reviewInput = node} />
            <button type="submit" className="z-depth-1" onClick={this.addReview}>Submit</button>
          </div>
        }
      </li>
    );
  }
}

AddReviewForm.propTypes = {
  dispatch: PropTypes.func.isRequired,
  user: PropTypes.shape({
    id: PropTypes.number.isRequired,
    fullname: PropTypes.string.isRequired,
  }).isRequired,
  id: PropTypes.number.isRequired
};

export default connect(null)(AddReviewForm);
