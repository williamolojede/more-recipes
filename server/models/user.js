import bcrypt from 'bcrypt';

module.exports = (sequelize, DataTypes) => {
  const User = sequelize.define('User', {
    username: {
      allowNull: true,
      unique: true,
      validate: {
        max: 15,
      },
      type: DataTypes.STRING,
    },
    email: {
      allowNull: false,
      unique: true,
      validate: {
        isEmail: true,
      },
      type: DataTypes.STRING,
    },
    fullname: {
      allowNull: false,
      validate: {
        max: 50,
      },
      type: DataTypes.STRING,
    },
    password: {
      allowNull: false,
      type: DataTypes.STRING,
      set(password) {
        const hash = bcrypt.hashSync(password, 10);
        this.setDataValue('password', hash);
      }
    },
  });

  User.associate = (models) => {
    User.hasMany(models.Recipe, {
      foreignKey: 'owner',
      as: 'recipes',
    });

    User.hasMany(models.Favorite, {
      foreignKey: 'userId',
      as: 'favorites',
    });

    User.hasMany(models.Review, {
      foreignKey: 'userId',
      as: 'reviews',
    });
  };

  User.authenticate = (email, password, callback) => {
    // console.log(email, password);
    User.findOne({
      where: { email },
    })
      .then((user) => {
        // check if it finds no user
        if (!user) {
          const err = new Error('User not found');
          err.status = 404;
          return callback(err);
        }
        const isMatch = bcrypt.compareSync(password, user.password);

        // check if password matches
        if (isMatch) return callback(null, user);
        return callback();
      })
      .catch(err => callback(err));
  };
  return User;
};
