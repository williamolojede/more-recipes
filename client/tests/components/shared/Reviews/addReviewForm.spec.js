import React from 'react';
import AddReviewForm from '../../../../components/shared/Reviews/AddReviewForm';
import user from '../../../__mocks__/user.mock';


function setup() {
  const props = {
    dispatch: jest.fn().mockReturnValue(Promise.resolve()),
    recipeId: 1,
    currentUser: user,
    limit: 6
  };

  const wrapper = mount(<AddReviewForm {...props} />);

  return {
    wrapper,
    props
  };
}

describe('Add review form ', () => {
  it('should render with the right amount of elements', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should dispatch an action if button is clicked and input field is not empty', () => {
    const { wrapper } = setup();
    const textarea = wrapper.find('textarea');
    const submitButton = wrapper.find('button');
    textarea.at(0).instance().value = 'test';

    const dispatchSpy = jest.spyOn(wrapper.instance().props, 'dispatch');
    // const addReviewSpy = jest.spyOn(wrapper.instance(), 'addReview');

    submitButton.simulate('click');
    expect(dispatchSpy).toHaveBeenCalledTimes(1);
  });

  it('should not dispatch an action if button is clicked and input field is empty', () => {
    const { wrapper } = setup();
    const textarea = wrapper.find('textarea');
    const submitButton = wrapper.find('button');
    textarea.at(0).instance().value = '';

    const dispatchSpy = jest.spyOn(wrapper.instance().props, 'dispatch');
    submitButton.simulate('click');
    expect(dispatchSpy).toHaveBeenCalledTimes(0);
  });
});
