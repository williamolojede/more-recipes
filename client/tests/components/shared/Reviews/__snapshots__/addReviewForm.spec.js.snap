// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add review form  should render with the right amount of elements 1`] = `
<AddReviewForm
  currentUser={
    Object {
      "fullname": "John Doe",
      "id": 1,
    }
  }
  dispatch={[Function]}
  limit={6}
  recipeId={1}
>
  <div
    className="review__card review__card-edit"
    id="write-review"
  >
    <div
      className="wrapper"
    >
      <UserImg
        type="inReview"
        user={
          Object {
            "fullname": "John Doe",
            "id": 1,
          }
        }
      >
        <div
          className="user-image inReview"
          data-for="username-inReview"
          data-tip="John Doe"
        >
          <div
            className="user-image__no-img"
          >
            <span>
              J
            </span>
            <span>
              D
            </span>
          </div>
        </div>
      </UserImg>
      <div
        className="review__card-edit__form"
      >
        <textarea
          placeholder="your review..."
        />
        <button
          className=" btn z-depth-1"
          onClick={[Function]}
          type="button"
        >
          Submit
        </button>
      </div>
    </div>
  </div>
</AddReviewForm>
`;
